### PR TITLE
Calculator

### PR DESCRIPTION
**Pull request by LeviOrtega**
## 📃 Description

Created the Calculator Sample app

## 🤔 Motivation and Context

To provide developers more examples in which the dual screen can be used on the surface duo

## 🧪 How Has This Been Tested?

Only manual testing of the application 

## 📷 Screenshots (if appropriate)

single_portrait

dual_portrait

dual_landscape

single_landscape

## 📦 Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Enhancement to a current functionality
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Bug fix (non-breaking change which fixes an issue. Please add issue link here too).

## ✅ Checklist

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] This PR contains new tests that cover the new code.
- [ ] This PR refactor previous tests to cover the new code.
- [ ] This PR is part of a set of PRs that build a bigger feature. If so, please, add here links to previously merged or open PRs.

### CODE REVIEW

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/MainPage.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import com.microsoft.device.dualscreen.twopanelayout.TwoPaneMode
+import com.microsoft.device.dualscreen.windowstate.WindowState
+
+val calculatorModel: CalculatorModel = CalculatorModel()
+val historyModel: HistoryModel = HistoryModel()
+
+@Composable
+fun MainPage(windowState: WindowState) {
+    TwoPaneLayout(
+        paneMode = TwoPaneMode.TwoPane,
```
Comment (khalp) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/MainPage.kt:16]: You can remove this line, `TwoPane` is already the default pane mode

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/MainPage.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        pane1 = {
+            BasicCalculatorPage(windowState = windowState)
+        },
+        pane2 = {
+            AdvancedCalculatorPage(windowState = windowState)
+        }
```
Comment (khalp) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/MainPage.kt:22]: just a style preference, but you can keep both of these to one line for better readability
```suggestion
        pane1 = { BasicCalculatorPage(windowState = windowState) },
        pane2 = { AdvancedCalculatorPage(windowState = windowState) }
```

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/BasicCalculatorPage.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            History(
+                addRecordToTop = false,
+                modifier = Modifier
+                    .fillMaxHeight(0.25f)
+                    .align(Alignment.Start)
+            )
+
+            Answer(modifier = Modifier.align(Alignment.End))
+
+            Row() {
```
Comment (khalp) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/BasicCalculatorPage.kt:40]: can remove the parentheses after `Row`

Comment (khalp) [reply Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/BasicCalculatorPage.kt:40]: same on line 53 😄

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/Enums.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    return when (eq) {
+        Equation.ADD -> stringResource(id = R.string.add)
+        Equation.SUB -> stringResource(id = R.string.subtract)
+        Equation.MUL -> stringResource(id = R.string.multiply)
+        Equation.DIV -> stringResource(id = R.string.divide)
+        Equation.POW -> stringResource(id = R.string.power)
+        Equation.MOD -> stringResource(id = R.string.modulus)
+        else -> {
+            ""
+        }
```
Comment (khalp) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/Enums.kt:29]: style comment but can simplify to one line as well
```suggestion
        else -> ""
```

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/Calculations.kt [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            )
+        ).toFloat().toString() else kotlin.math.sin(calculatorModel.x.toDouble()).toFloat().toString()
+    } else {
+        calculatorModel.y = if (calculatorModel.degrees) kotlin.math.sin(
+            Math.toRadians(
+                calculatorModel.y.toDouble()
+            )
+        ).toFloat().toString() else kotlin.math.sin(calculatorModel.y.toDouble()).toFloat().toString()
+    }
+}
```
Comment (khalp) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/Calculations.kt:43]: optional style suggestion to improve readability/remove duplicate code (if you like this suggestion, you can probably do something similar for the other calculations that require similar conversions):
```suggestion
fun sin() {
    if (calculatorModel.isOnX) {
        calculatorModel.x = executeTrigFunction(::sin, calculatorModel.x)
    } else {
        calculatorModel.y = executeTrigFunction(::sin, calculatorModel.y)
    }
}

private fun executeTrigFunction(trigFunction: (Double) -> Double, rawInput: String): String {
    val radiansInput = if (calculatorModel.degrees) Math.toRadians(rawInput.toDouble()) else rawInput.toDouble()

    return trigFunction(radiansInput).toFloat().toString()
}
```

Note: for the suggestion above, had to add `import kotlin.math.sin`

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/CalculatorModel.kt [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            Equation.ADD -> add()
+            Equation.SUB -> sub()
+            Equation.MUL -> mul()
+            Equation.DIV -> div()
+            Equation.POW -> power()
+            Equation.MOD -> modulus()
+            else -> {}
+        }
+        decimalReformat()
+        if (currentEquation != Equation.NONE) historyModel.addToRecords(Record(prevX, prevY, currentEquation, this.x))
```
Comment (khalp) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/CalculatorModel.kt:32]: I would separate this to two lines for better readability - I think in general the main reason to keep if statements on one line if is you're using them for assignments (ex: `val example = if (flag) "flag is true" else "flag is false"`)

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/CalculatorModel.kt [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            Action.LOG -> log()
+            Action.LN -> ln()
+            Action.E -> e()
+            Action.PI -> pi()
+            Action.SQRT -> squareRoot()
+            Action.SQRD -> squared()
+            Action.DEG -> toggleDegree()
+            Action.PRCT -> percent()
+            Action.SIGN -> toggleNegative()
+            else -> {}
```
Comment (khalp) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/CalculatorModel.kt:56]: I think you can delete this

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/CalculatorModel.kt [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+        if (floatY == 0.0f) { this.y = "0" } else if ((floatY - floatY.roundToInt()) == 0.0f) { this.y = (floatY.roundToInt()).toString() }
+    }
+
+    private fun clearValues() {
+        // if the user clears when x is 0, we will do an all clear
+        if (x == "0" && isOnX) { historyModel.clearRecords() }
+        resetValues()
+    }
+    private fun resetToX() {
```
Comment (khalp) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/CalculatorModel.kt:80]: style/readability - it's good to have spaces in between functions (if you ctrl + alt + L in all your files, it'll fix spacing issues for you)

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/components/Answer.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+fun Answer(modifier: Modifier = Modifier) {
+    val eq: String = if (!calculatorModel.isOnX) " " + equationStringEquivalent(calculatorModel.currentEquation) + " " else ""
+    val y: String = if (!calculatorModel.isOnX) calculatorModel.y else ""
+    LazyColumn(
+        modifier = modifier
+            .padding(
+                start = 20.dp,
+                top = 0.dp,
+                end = 20.dp,
+                bottom = 0.dp
```
Comment (khalp) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/components/Answer.kt:25]: Since the default value for these padding values is already 0.dp, you can delete the top/bottom parameters
```suggestion
                start = 20.dp,
                end = 20.dp,
```

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/components/Buttons.kt [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+import com.microsoft.device.display.samples.calculator.models.equationStringEquivalent
+import com.microsoft.device.display.samples.calculator.ui.pages.calculatorModel
+
+@Composable
+fun EquationButton(eq: Equation, color: Color = MaterialTheme.colors.primary) {
+    ButtonLayout(
+        color = color,
+        content = equationStringEquivalent(eq)
+    ) { calculatorModel.updateEquation(eq) }
+}
```
Comment (khalp) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/ui/pages/components/Buttons.kt:36]: optional style suggestion: since this lambda is for an onClick method and not composable content, it may be more clear to explicitly name the parameter
```suggestion
@Composable
fun EquationButton(eq: Equation, color: Color = MaterialTheme.colors.primary) {
    ButtonLayout(
        color = color,
        content = equationStringEquivalent(eq)
        onClick = { calculatorModel.updateEquation(eq) }
    ) 
}
```

**Review by khalp — COMMENTED** [review]
This looks awesome Levi, great job!! Your Compose code looks great and I love the themed colors 😄 

Most of my comments were just optional style suggestions, and for future reference, if you ever want to learn about unit testing/UI testing, I think this would be a great sample to add those to!

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/Calculations.kt (hunk 5 above)
Comment (LeviOrtega) [reply Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/Calculations.kt:43]: Thank you!! I was trying to figure out how to do something like this but couldn't figure it out.  Appreciate it

File: Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/Calculations.kt [hunk 11]
```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+package com.microsoft.device.display.samples.calculator.models
```
Comment (CesarValiente) [inline Calculator/src/main/java/com/microsoft/device/display/samples/calculator/models/Calculations.kt:6]: I think you may reconsider how to use the math operations below. In all functions there is not an input and output, it takes its input and writes its output in an external model (think about this as an external variable). All these functions cannot be unit tested since you can not provide specific input and check the result as an output. This is easier to see when implementing tests, but basically, you cannot test, let say, `add` function, because you don't have a way to provide the set of numbers you want to use in the add operation and check what the function returns as its result, because the function is nor returning anything.

The correct way would be:
```
fun add(param1: Int, param2: int): Int {
return param1 + param2 
}
```
Then when testing you could do:
```
@Test
fun should_my_add_function_do_the_math_correctly() {
    val expected = 3
    val result = add (1, 2)
    assertThat (result, is(expected))
}
As a good rule to follow, try always to avoid global objects that can be called along the whole app and modified everywhere since it will be difficult/impossible to test + there will be many side effects that you cannot control and give you problems.

**Comment by CesarValiente** [comment]
Woaw! Fantastic work Levi! 👏👏👏

Agree with Kirsten, this apps looks like a fantastic place to learn about implementing tests (and is also linked to my comment):
- Adding unittests to the custom math operations (not to math functions that the language brings since there are tested already, but on any combination of them or complex functions you have created). While implementing unit tests you will see how what I've mentioned above makes sense and see how the current implementation cannot be unit tested.
- Adding uitests so you can, for example, create tests that tap in the num-buttons, do operations, and check that the result is the expected in the result textview.

No worries, we will learn all of that together :-)

**Review by khalp — APPROVED** [review]
Thanks for addressing my feedback, looks great!

File: Calculator/build.gradle [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    id 'com.android.application'
+    id 'org.jetbrains.kotlin.android'
+}
+
+android {
+    compileSdk 32
+
+    defaultConfig {
+        applicationId "com.microsoft.device.display.samples.calculator"
+        minSdk 23
```
Comment (joyl1216) [inline Calculator/build.gradle:16]: I think you can use the root project configuration for these. 
```
        minSdkVersion rootProject.ext.minSdkVersion
        targetSdkVersion rootProject.ext.targetSdkVersion
        testInstrumentationRunner rootProject.ext.config.testInstrumentationRunner

File: Calculator/README.md [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+This sample is built with Jetpack Compose, the new UI framework in Android.
+
+## Getting Started
+
+To learn how to load apps on the Surface Duo emulator, see the [documentation](https://docs.microsoft.com/dual-screen/android), and follow [the blog](https://devblogs.microsoft.com/surface-duo).
+
+## Features
+
+The sample uses the [Companion Pane](https://docs.microsoft.com/dual-screen/introduction#companion-pane) app pattern to show a basic Calculator on one pane, and additional more advanced grid of equations on the other pane. This sample also has a ticker tape, which is a history of entered equations that can be accessed later. Depending on the orientation, this ticker tape will be on the first or second pane.
```
Comment (joyl1216) [inline Calculator/README.md:11]: I may argue that this sample is using `Extended Canvas`, rather than `Companion Pane`.

**Comment by joyl1216** [comment]
@LeviOrtega Thanks for your excellent work! I would suggest pre-installing this Calculator into our emulator. The design and the app look very good, especially in the dual-screen mode.
To do that, we should get an app icon and maybe a new app name, to emphasize the `dual` or `two`. You can reach out to our designer Yuran Liu.

And I agree with Kristen and Cesar, that the sample is a good candidate to implement the testing, which can make sure all the math logic works as expected.

File: Calculator/src/androidTest/java/com/microsoft/device/display/samples/calculator/CalculatorTest.kt [hunk 14]
```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+package com.microsoft.device.display.samples.calculator
```
[the remark this answers is not in the document]
Comment (LeviOrtega) [reply Calculator/src/androidTest/java/com/microsoft/device/display/samples/calculator/CalculatorTest.kt:6]: Ah I see, sounds good. Thank you

**Comment by joyl1216** [comment]
@LeviOrtega `0 mod 0 =` would crash the app currently. May I please suggest adding some edge cases to the test?

File: DyAdd/src/main/res/values/strings.xml [hunk 15]
```diff
@@ -0,0 +1,32 @@
+<!--
+  ~ Copyright (c) Microsoft Corporation. All rights reserved.
+  ~ Licensed under the MIT License.
+  -->
+
+<resources>
+    <string name="app_name">Calculator</string>
```
Comment (khalp) [inline DyAdd/src/main/res/values/strings.xml:7]: You should change this to DyAdd as well so the app name on the device is updated!

**Comment by khalp** [comment]
You also need to update the code to make sure the icon is being used:
image

**Comment by joyl1216** [comment]
@LeviOrtega Good job! After we merge it, I will preload it into our emulator.
